pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1841,346 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
+		validBus := false
+		for _, supportedBus := range supportedBuses {
+			if bus == supportedBus {
+				validBus = true
 			}
 		}
-
-		// Verify serial number is made up of valid characters for libvirt, if provided
-		if disk.Serial != "" && !isValidExpression(disk.Serial) {
+		if !validBus {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s must be made up of the following characters [A-Za-z0-9_.+-], if specified", field.Index(idx).String()),
-				Field:   field.Index(idx).Child("serial").String(),
+				Message: fmt.Sprintf("%s is set with an unrecognized bus %s, must be one of: %v", field.Index(idx).String(), bus, supportedBuses),
+				Field:   field.Index(idx).Child(diskType, "bus").String(),
 			})
 		}
+	}
 
-		// Verify serial number is within valid length, if provided
-		if disk.Serial != "" && len([]rune(disk.Serial)) > maxStrLen {
-			causes = append(causes, metav1.StatusCause{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s must be less than or equal to %d in length, if specified", field.Index(idx).String(), maxStrLen),
-				Field:   field.Index(idx).Child("serial").String(),
-			})
-		}
+	// special case. virtio is incompatible with CD-ROM for q35 machine types
+	if diskType == "cdrom" && bus == v1.DiskBusVirtio {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("Bus type %s is invalid for CD-ROM device", bus),
+			Field:   field.Index(idx).Child("cdrom", "bus").String(),
+		})
+	}
+	// sata disks (in contrast to sata cdroms) don't support readOnly
+	if disk.Disk != nil && bus == v1.DiskBusSATA && disk.Disk.ReadOnly {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s hard-disks do not support read-only.", bus),
+			Field:   field.Index(idx).Child("disk", "bus").String(),
+		})
+	}
 
-		// Verify if cache mode is valid
-		if disk.Cache != "" && disk.Cache != v1.CacheNone && disk.Cache != v1.CacheWriteThrough && disk.Cache != v1.CacheWriteBack {
-			causes = append(causes, metav1.StatusCause{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s has invalid value %s", field.Index(idx).Child("cache").String(), disk.Cache),
-				Field:   field.Index(idx).Child("cache").String(),
-			})
-		}
+	// Reject defining DedicatedIOThread to a disk without VirtIO bus since this configuration
+	// is not supported in libvirt.
+	isIOThreadsWithUnsupportedBus := disk.DedicatedIOThread != nil && *disk.DedicatedIOThread && bus != v1.DiskBusVirtio
+	if isIOThreadsWithUnsupportedBus {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueNotSupported,
+			Message: fmt.Sprintf("IOThreads are not supported for disks on a %s bus", bus),
+			Field:   field.Child("domain", "devices", "disks").Index(idx).String(),
+		})
+	}
```

Comment:
I would rearrange this in order to avoid to loop over the bus types. You can have a a switch based on the bus type and add the various causes inside the case of each bus type. Something like:
```golang
        switch bus {
        case "ide":
                causes = append(causes, metav1.StatusCause{
                        Type:    metav1.CauseTypeFieldValueInvalid,
                        Message: "IDE bus is not supported",
                        Field:   field.Index(idx).Child(diskType, "bus").String(),
                })
        case v1.DiskBusVirtio:
                if diskType == "cdrom" {
                        causes = append(causes, metav1.StatusCause{
                                Type:    metav1.CauseTypeFieldValueInvalid,
                                Message: fmt.Sprintf("Bus type %s is invalid for CD-ROM device", bus),
                                Field:   field.Index(idx).Child("cdrom", "bus").String(),
                        })
                }
          [...]
        case v1.DiskBusSATA:
                if disk.Disk != nil && disk.Disk.ReadOnly {  ....  }
        case v1.DiskBusSCSI, v1.DiskBusUS:
                break
        default:
                causes = append(causes, metav1.StatusCause{
                        Type:    metav1.CauseTypeFieldValueInvalid,
                        Message: fmt.Sprintf("%s is set with an unrecognized bus %s, must be one of: %v", field.Index(idx).String(), bus, supportedBuses),
                        Field:   field.Index(idx).Child(diskType, "bus").String(),
                })
        }
```